log/packetlogger.go

```diff
@@ -0,0 +1,155 @@
+func WithEnableRollbar(enable bool) LoggerOption {
+	return func(args *PacketLogger) { args.EnableRollbar = enable }
+}
+
+// WithRollbarConfig customizes the Rollbar details
+func WithRollbarConfig(config RollbarConfig) LoggerOption {
+	return func(args *PacketLogger) { args.RollbarConfig = config }
+}
+
+// PacketLogger is a wrapper around zap.SugaredLogger
+type PacketLogger struct {
+	logr.Logger
+	LogLevel              string
+	OutputPaths           []string
+	ServiceName           string
+	KeysAndValues         map[string]interface{}
+	EnableErrLogsToStderr bool
+	EnableRollbar         bool
+	RollbarConfig         RollbarConfig
+}
+
+// LoggerOption for setting optional values
+type LoggerOption func(*PacketLogger)
+
+// NewPacketLogger is the opionated packet logger setup
+func NewPacketLogger(opts ...LoggerOption) (logr.Logger, *zap.Logger, error) {
+	// defaults
+	const (
+		defaultLogLevel    = "info"
+		defaultServiceName = "not/set"
+	)
+	var (
+		defaultOutputPaths   = []string{"stdout"}
+		defaultKeysAndValues = map[string]interface{}{"service": defaultServiceName}
+		zapConfig            = zap.NewProductionConfig()
+		zLevel               = zap.InfoLevel
+		rollbarOptions       zap.Option
+		defaultRollbarConfig = RollbarConfig{
+			Token:   "123",
+			Env:     "production",
+			Version: "1",
+		}
+	)
+
+	pl := &PacketLogger{
+		Logger:        nil,
+		LogLevel:      defaultLogLevel,
+		OutputPaths:   defaultOutputPaths,
+		ServiceName:   defaultServiceName,
+		KeysAndValues: defaultKeysAndValues,
+		EnableRollbar: false,
+		RollbarConfig: defaultRollbarConfig,
+	}
+
+	for _, opt := range opts {
+		opt(pl)
+	}
+
+	switch pl.LogLevel {
+	case "debug":
+		zLevel = zap.DebugLevel
+	}
+	zapConfig.Level = zap.NewAtomicLevelAt(zLevel)
+	zapConfig.OutputPaths = pl.OutputPaths
+	zapConfig.OutputPaths = sliceDedupe(append(zapConfig.OutputPaths, "stdout"))
+	zapConfig.InitialFields = pl.KeysAndValues
+
+	zapLogger, err := zapConfig.Build()
+	if err != nil {
+		return pl, zapLogger, errors.Wrap(err, "failed to build logger config")
+	}
+
+	if pl.EnableRollbar {
+		rollbarOptions = pl.RollbarConfig.setupRollbar(pl.ServiceName, zapLogger)
+		zapLogger = zapLogger.WithOptions(rollbarOptions)
+	}
+	if pl.EnableErrLogsToStderr {
+		zapLogger = zapLogger.WithOptions(errLogsToStderr(zapConfig))
+	}
+
+	pl.Logger = zapr.NewLogger(zapLogger)
+	return pl, zapLogger, err
+}
+
+func sliceDedupe(elements []string) []string {
+	encountered := map[string]bool{}
+	result := []string{}
+
+	for v := range elements {
+		if encountered[elements[v]] {
+		} else {
+			encountered[elements[v]] = true
+			result = append(result, elements[v])
+		}
+	}
+	return result
+}
+
+func errLogsToStderr(c zap.Config) zap.Option {
+	errorLogs := zap.LevelEnablerFunc(func(lvl zapcore.Level) bool {
+		return lvl >= zapcore.ErrorLevel
+	})
+	nonErrorLogs := zap.LevelEnablerFunc(func(lvl zapcore.Level) bool {
+		return lvl < zapcore.ErrorLevel
+	})
```

Comment:
```go
nonErrorLogs := zap.LevelEnablerFunc(func(lvl zapcore.Level) bool {
		return !errorLogs(lvl)
})
```
this way we can't *ever* get it wrong by typo somehow.

Reply:
updated